File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
#         l = 0
#         r = 0
#         res = 0
        
#         while r < len(prices):
#             profit = prices[r] - prices[l]
#             res = max(res, profit)
#             if profit < 0:
#                 l = r
#             r+=1
#         return res
            
    
        min_profit = prices[0]
        max_profit = 0
        
        for i in range(1,len(prices)):
            profit = prices[i] - min_profit
            
            if profit < 0:
                min_profit = prices[i]
            else:
                max_profit = max(max_profit, profit)
                
        return max_profit

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int max_profit = 0;
        int buy_price = prices[0];
        for(int i = 1; i < prices.length; i++){
            if (buy_price > prices[i]){
                buy_price = prices[i];
            }
            max_profit = Math.max(max_profit, prices[i] - buy_price);
        }

        return max_profit;
    }
}